### ISSUE TITLE
New rule: side effects inside pure lambdas

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#7764]
## Expected Behavior of the rule
The rule should detect code inside lambdas that should be pure functions that contain side effects. For example:

```kotlin
listOf(1, 2)
  .map {
    println(it)
    it + 1
  }
```

The println shouldn't be there. This code should be:

```kotlin
listOf(1, 2)
  .onEach { println(it) } 
  .map { it + 1 }
```

This way you separate impure functions like onEach to pure functions like map. 

## Context
When instead of println you have a more complex code the benefits are way higher. 

There are a lot of cases on the standard library where this happens, `map` is just an example: `filter`, `flatMap`... 

This rule is not going to be easy to implement. And the rule could have false negatives in cases where it can't know if something is a side effect or not.

The first step would be to find if a block contributes to the output of a lambda. If it doesn't contributes it should be moved to an onEach.

**Comment by cortinico** [comment]
> that contain side effects

I believe this would be the trickiest part. How do you detect that a function is having side effects or not? You could have:

```kotlin
listOf(1, 2)
  .map {
    it + computeValueAndHaveSideEffects(1)
  }
```

**Comment by BraisGabin** [comment]
Something like that wouldn't be flagged. Only code inside the lambda that doesn't contribute to the return value. We should need to assume that the value it on map is always immutable.

We can later expand the rule but I think that for now that would be good enough. Controversial rule for sure but I think that it will help with the readability of the code.
